commit_test.go

```diff
@@ -62,6 +63,47 @@ func (s *SuiteCommit) TestParents(c *C) {
 	c.Assert(output, DeepEquals, expected)
 }
 
+func (s *SuiteCommit) TestCommitEncodeDecodeIdempotent(c *C) {
+	ts, err := time.Parse(time.RFC3339, "2006-01-02T15:04:05-07:00")
+	c.Assert(err, IsNil)
+	commits := []*Commit{
+		&Commit{
+			Author:    Signature{Name: "Foo", Email: "foo@example.local", When: ts},
+			Committer: Signature{Name: "Bar", Email: "bar@example.local", When: ts},
+			Message:   "Message\n\nFoo\nBar\nBaz\n\n",
+			tree:      core.NewHash("b029517f6300c2da0f4b651b8642506cd6aaf45d"),
+			parents:   []core.Hash{core.NewHash("b029517f6300c2da0f4b651b8642506cd6aaf45d")},
+		},
+		&Commit{
+			Author:    Signature{Name: "Foo", Email: "foo@example.local", When: ts},
+			Committer: Signature{Name: "Bar", Email: "bar@example.local", When: ts},
+			Message:   "Message\n\nFoo\nBar\nBaz\n\n",
+			tree:      core.NewHash("b029517f6300c2da0f4b651b8642506cd6aaf45d"),
+			parents: []core.Hash{
+				core.NewHash("b029517f6300c2da0f4b651b8642506cd6aaf45d"),
+				core.NewHash("c029517f6300c2da0f4b651b8642506cd6aaf45d"),
+				core.NewHash("d029517f6300c2da0f4b651b8642506cd6aaf45d"),
+				core.NewHash("e029517f6300c2da0f4b651b8642506cd6aaf45d"),
+			},
+		},
+	}
+	for _, commit := range commits {
+		// Compute hash
+		obj := &core.MemoryObject{}
+		err = commit.Encode(obj)
+		c.Assert(err, IsNil)
+		commit.Hash = obj.Hash()
+
+		obj = &core.MemoryObject{}
+		err = commit.Encode(obj)
+		c.Assert(err, IsNil)
+		newCommit := &Commit{}
+		err = newCommit.Decode(obj)
+		c.Assert(err, IsNil)
+		c.Assert(newCommit, DeepEquals, commit)
+	}
+}
+
```

Comment:
Yes. I used to this this with a git repository fixture, doing Decode first and then Encode. Changing it to mocks + Encode + Decode requires us to calculate the hash of the encoded object to know what should we expect.

I changed it to be a bit more concise though.